rust/src/state_machine/phases/mod.rs

```diff
@@ -35,7 +35,45 @@ use tokio::sync::oneshot;
 #[async_trait]
 pub trait Phase<R> {
     /// Moves from this state to the next state.
-    async fn next(mut self) -> Option<StateMachine<R>>;
+    fn next(self) -> Option<StateMachine<R>>;
+
+    /// Run this phase to completion
+    async fn run(&mut self) -> Result<(), StateError>;
+
+    /// Return `true` if this is the error phase
```

Comment:
I think we can add a `PhaseStates` enum an add a `phase` field to the CoordinatorState. It's somewhat redundant, of course, but we may need this information (in which phase the coodrinator was in) later if we want to restore the coordinator state.

Reply:
You'd prefer that to the current `is_xxx` methods? As you said it'd be redundant so I'm not really convinced.

Reply:
difficult to decide. I think it would be a bit cleaner, but on the other hand, we can already provide information about the state of the state machine without introducing an enum. 
I think using the `is_xxx` methods is fine

Reply:
Let's see if @finiteprods and @janpetschexain have an opinion on that matter.

Reply:
i personally would take an enum, but the `is_x` state functions are ok as well.

Reply:
You guys were right: it looks much better with 859c5b0

Reply:
Looks nice👍